// @kaltura-ng/kaltura-logger/esm5/lib/kaltura-logger.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { KalturaLogger, KalturaLoggerName } from './kaltura-logger.service';
import { KalturaLoggerRecordService } from './kaltura-logger-record.service';
if (window && window.onerror) {
    window.onerror = null;
}
var KalturaLoggerModule = /** @class */ (function () {
    function KalturaLoggerModule() {
    }
    KalturaLoggerModule_1 = KalturaLoggerModule;
    KalturaLoggerModule.forRoot = function (name) {
        return {
            ngModule: KalturaLoggerModule_1,
            providers: [
                KalturaLogger,
                {
                    provide: KalturaLoggerName, useValue: name
                },
                KalturaLoggerRecordService
            ]
        };
    };
    var KalturaLoggerModule_1;
    KalturaLoggerModule = KalturaLoggerModule_1 = __decorate([
        NgModule({
            imports: [
                CommonModule
            ],
            declarations: [],
            exports: [],
            providers: []
        })
    ], KalturaLoggerModule);
    return KalturaLoggerModule;
}());
export { KalturaLoggerModule };
