// @kaltura-ng/kaltura-common/esm2015/lib/kaltura-common.module.js
var KalturaCommonModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { KalturaUtils } from './utils/kaltura-utils';
import { APP_STORAGE_TOKEN, AppStorage } from './app-storage.service';
import { EmptyLogger, KalturaLoggerInjectionToken } from './kaltura-logger';
let KalturaCommonModule = KalturaCommonModule_1 = class KalturaCommonModule {
    // constructor(@Optional() @SkipSelf() module : KalturaCoreModule, private appBootstrap : AppBootstrap)
    // {
    //     if (module) {
    //         throw new Error("KMCngCoreModule module imported twice.");
    //     }
    // }
    static forRoot() {
        return {
            ngModule: KalturaCommonModule_1,
            providers: [
                { provide: APP_STORAGE_TOKEN, useClass: AppStorage },
                KalturaUtils,
                { provide: KalturaLoggerInjectionToken, useClass: EmptyLogger }
            ]
        };
    }
};
KalturaCommonModule = KalturaCommonModule_1 = __decorate([
    NgModule({
        imports: [
            CommonModule,
        ],
        declarations: [],
        exports: [],
        providers: []
    })
], KalturaCommonModule);
export { KalturaCommonModule };
