// @kaltura-ng/kaltura-ui/esm2015/lib/sticky/components/sticky.component.js
import { __decorate, __metadata } from "tslib";
import { Component, ElementRef, Input, AfterViewInit, OnDestroy, ViewChild, Renderer2 } from '@angular/core';
import { StickyScrollService } from '../services/sticky-scroll.service';
import { StickyDirective } from '../directives/sticky.directive';
import { cancelOnDestroy } from '@kaltura-ng/kaltura-common';
let StickyComponent = class StickyComponent {
    constructor(_stickyScrollService, _renderer) {
        this._stickyScrollService = _stickyScrollService;
        this._renderer = _renderer;
        this.wrapperHeight = 0;
        this.scrollOffset = 0;
    }
    ngAfterViewInit() {
        setTimeout(() => {
            this._updateLayout();
        }, 0);
        this._stickyScrollService.resizeStatus$.pipe(cancelOnDestroy(this)).subscribe(event => {
            this._updateDimensions();
        });
    }
    _updateLayout() {
        if (this.content.nativeElement.children.length > 0) {
            this.wrapperHeight = this.content.nativeElement.children[0].clientHeight;
        }
        else {
            console.warn("Sticky component::could not access content.");
        }
    }
    updateLayout(manual = true) {
        setTimeout(() => {
            this._updateLayout();
            this._sticky.update();
        }, 0);
    }
    _updateDimensions() {
        this._renderer.setStyle(this.stickyDiv.nativeElement, 'width', `${this.content.nativeElement.offsetWidth}px`);
    }
    ngOnDestroy() {
    }
};
StickyComponent.ctorParameters = () => [
    { type: StickyScrollService },
    { type: Renderer2 }
];
__decorate([
    Input(),
    __metadata("design:type", Object)
], StickyComponent.prototype, "scrollOffset", void 0);
__decorate([
    Input('stickyClass'),
    __metadata("design:type", String)
], StickyComponent.prototype, "stickyClass", void 0);
__decorate([
    Input('container'),
    __metadata("design:type", Object)
], StickyComponent.prototype, "container", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], StickyComponent.prototype, "stickyId", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], StickyComponent.prototype, "sticksTo", void 0);
__decorate([
    ViewChild('contentWrapper', { static: true }),
    __metadata("design:type", ElementRef)
], StickyComponent.prototype, "content", void 0);
__decorate([
    ViewChild('stickyDiv', { static: true }),
    __metadata("design:type", ElementRef)
], StickyComponent.prototype, "stickyDiv", void 0);
__decorate([
    ViewChild(StickyDirective, { static: true }),
    __metadata("design:type", StickyDirective)
], StickyComponent.prototype, "_sticky", void 0);
StickyComponent = __decorate([
    Component({
        selector: 'kSticky',
        template: "<div class=\"kStickyWrapper\" [style.height]=\"wrapperHeight + 'px'\" #contentWrapper>\n    <div #stickyDiv class=\"kSticky\" kSticky [scrollOffset]=\"scrollOffset\" [sticksTo]=\"sticksTo\" [stickyClass]=\"stickyClass\" [stickyId]=\"stickyId\" (onStickyEvent)=\"_updateDimensions()\" (onUnStickyEvent)=\"_updateDimensions()\">\n        <ng-content></ng-content>\n    </div>\n</div>\n",
        styles: [".kStickyWrapper{display:block}.kStickyWrapper .kSticky{z-index:501}"]
    }),
    __metadata("design:paramtypes", [StickyScrollService, Renderer2])
], StickyComponent);
export { StickyComponent };
