// @kaltura-ng/kaltura-ui/esm2015/lib/sticky/services/sticky-scroll.service.js
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs/Subject';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
let StickyScrollService = class StickyScrollService {
    constructor() {
        this._scrollSubject = new Subject();
        this.scrollStatus$ = this._scrollSubject.asObservable();
        this._resizeSubject = new Subject();
        this.resizeStatus$ = this._resizeSubject.asObservable();
        this._layoutSubject = new BehaviorSubject({});
        this.layoutSubject$ = this._layoutSubject.asObservable();
        this.stickyElements = {};
        this.manageScrollFunc = this.manageScrollEvent.bind(this);
        this.manageResizeFunc = this.manageResizeEvent.bind(this);
        window.addEventListener('scroll', this.manageScrollFunc);
        window.addEventListener('resize', this.manageResizeFunc);
    }
    ngOnDestroy() {
        window.removeEventListener('scroll', this.manageScrollFunc);
        window.removeEventListener('resize', this.manageResizeFunc);
    }
    manageScrollEvent() {
        this._scrollSubject.next({});
    }
    manageResizeEvent() {
        this._resizeSubject.next({});
    }
    attach(id) {
        if (!id) {
            throw new Error("sticky service::missing id on attach!");
        }
        else if (typeof this.stickyElements[id] !== "undefined") {
            throw new Error("sticky service::id already exists! (" + id + ")");
        }
        else {
            this.stickyElements[id] = null;
        }
    }
    detach(id) {
        if (!id) {
            console.warn("sticky service::missing id on attach!");
        }
        else {
            delete this.stickyElements[id];
        }
    }
    update(id, height, offset) {
        if (!id) {
            console.warn("sticky service::missing id!");
        }
        else {
            const currentValue = this.stickyElements[id];
            if (typeof currentValue !== "undefined") {
                if (currentValue === null || currentValue.height !== height || currentValue.offset !== offset) {
                    this.stickyElements[id] = { height, offset };
                    this._layoutSubject.next(this.stickyElements);
                }
            }
            else {
                throw new Error(`unknown sticky element '${id}' (did you attach it before updating?)`);
            }
        }
    }
};
StickyScrollService = __decorate([
    Injectable(),
    __metadata("design:paramtypes", [])
], StickyScrollService);
export { StickyScrollService };
