// @kaltura-ng/kaltura-ui/esm2015/lib/input-autofocus.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, Renderer2, Input, AfterViewInit } from '@angular/core';
let AutofocusDirective = class AutofocusDirective {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this._autofocus = true;
    }
    ngAfterViewInit() {
        if (this._autofocus) {
            setTimeout(() => {
                this.renderer.selectRootElement(this.el.nativeElement).focus();
            });
        }
    }
    set inputAutofocus(allowed) {
        this._autofocus = allowed;
    }
};
AutofocusDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    Input(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], AutofocusDirective.prototype, "inputAutofocus", null);
AutofocusDirective = __decorate([
    Directive({
        selector: '[kAutofocus]'
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2])
], AutofocusDirective);
export { AutofocusDirective };
