// @kaltura-ng/kaltura-ui/esm5/lib/input-helper/input-helper.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input } from '@angular/core';
var InputHelperComponent = /** @class */ (function () {
    function InputHelperComponent() {
        this.trigger = 'hover';
        this.triggerIcon = 'kIconhelp_full';
        this.icon = 'kIconhelp';
        this.width = 300;
        this.placement = { x: 'right', y: 'top' };
    }
    InputHelperComponent.prototype.ngOnInit = function () { };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], InputHelperComponent.prototype, "title", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], InputHelperComponent.prototype, "trigger", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], InputHelperComponent.prototype, "triggerIcon", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], InputHelperComponent.prototype, "icon", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], InputHelperComponent.prototype, "width", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], InputHelperComponent.prototype, "placement", void 0);
    InputHelperComponent = __decorate([
        Component({
            selector: 'kInputHelper',
            template: "<div class=\"kInputHelper\" #inputHelperContainer>\n    <a class=\"kIconhelpLink\" [class.kActive]=\"helperPopup.isShow\" #helperLink><i class=\"{{triggerIcon}}\"></i></a>\n\n    <kPopupWidget\n                  #helperPopup\n                  class=\"kInputHelperPopup\"\n                  [appendTo]=\"inputHelperContainer\"\n                  [targetOffset]=\"{x: 0, y: 20}\"\n                  [popupWidth]=\"width\"\n                  [closeBtn]=\"false\"\n                  [trigger]=\"trigger\"\n                  [placement]=\"placement\"\n                  [targetRef]=\"helperLink\">\n\n        <ng-template>\n            <div class=\"kInputHelperContainer kAverageElevation\">\n                <div *ngIf=\"title\" class=\"kInputHelperPopupHeader\">\n                    <i class=\"kIcon {{icon}}\"></i>\n                    {{title}}\n                </div>\n                <div class=\"kInputHelperPopupContent\">\n                    <ng-content></ng-content>\n                </div>\n            </div>\n        </ng-template>\n\n    </kPopupWidget>\n</div>\n",
            styles: [".kInputHelper{display:inline-block;vertical-align:middle;margin-left:6px;position:relative}.kInputHelper .kIconhelpLink{color:#999;text-decoration:none;position:relative;z-index:2}.kInputHelper .kIconhelpLink.kActive,.kInputHelper .kIconhelpLink:hover{color:#01accd}.kInputHelperPopup{border-radius:3px}.kInputHelperPopup ::ng-deep .kPopupWidget{box-shadow:none}.kInputHelperPopup .kInputHelperContainer{font-size:15px;line-height:20px;color:#999;padding:16px;border-radius:3px}.kInputHelperPopup .kInputHelperPopupHeader{color:#333;display:flex;align-items:center;padding-bottom:10px}.kInputHelperPopup .kInputHelperPopupHeader .kIcon{color:#999;font-size:24px;margin-right:14px}"]
        }),
        __metadata("design:paramtypes", [])
    ], InputHelperComponent);
    return InputHelperComponent;
}());
export { InputHelperComponent };
