// @kaltura-ng/kaltura-common/esm5/lib/operation-tag/operation-tag.module.js
import { __decorate, __metadata, __param } from "tslib";
import { Self, Optional, NgModule } from '@angular/core';
import { OperationTagStoreMediator } from "./operation-tag-store-mediator";
import { OperationTagManagerService } from "./operation-tag-manager.service";
var OperationTagModule = /** @class */ (function () {
    // Prevents angular from creating another instance for operationTagManagerService when defined as provider in more than one place
    // (Each component will get the same instance as in the module/component that called the operationTagManagerService.forRoot)
    function OperationTagModule(operationTagManagerService) {
        if (operationTagManagerService) {
            OperationTagStoreMediator.register(operationTagManagerService);
        }
    }
    OperationTagModule_1 = OperationTagModule;
    OperationTagModule.forRoot = function () {
        return {
            ngModule: OperationTagModule_1,
            providers: [
                OperationTagManagerService
            ]
        };
    };
    var OperationTagModule_1;
    OperationTagModule.ctorParameters = function () { return [
        { type: OperationTagManagerService, decorators: [{ type: Self }, { type: Optional }] }
    ]; };
    OperationTagModule = OperationTagModule_1 = __decorate([
        NgModule({
            imports: [],
            declarations: [],
            exports: [],
            providers: []
        }),
        __param(0, Self()), __param(0, Optional()),
        __metadata("design:paramtypes", [OperationTagManagerService])
    ], OperationTagModule);
    return OperationTagModule;
}());
export { OperationTagModule };
