// @kaltura-ng/kaltura-logger/esm2015/lib/kaltura-logger.module.js
var KalturaLoggerModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { KalturaLogger, KalturaLoggerName } from './kaltura-logger.service';
import { KalturaLoggerRecordService } from './kaltura-logger-record.service';
if (window && window.onerror) {
    window.onerror = null;
}
let KalturaLoggerModule = KalturaLoggerModule_1 = class KalturaLoggerModule {
    static forRoot(name) {
        return {
            ngModule: KalturaLoggerModule_1,
            providers: [
                KalturaLogger,
                {
                    provide: KalturaLoggerName, useValue: name
                },
                KalturaLoggerRecordService
            ]
        };
    }
};
KalturaLoggerModule = KalturaLoggerModule_1 = __decorate([
    NgModule({
        imports: [
            CommonModule
        ],
        declarations: [],
        exports: [],
        providers: []
    })
], KalturaLoggerModule);
export { KalturaLoggerModule };
