// @kaltura-ng/kaltura-common/esm5/lib/kaltura-common.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { KalturaUtils } from './utils/kaltura-utils';
import { APP_STORAGE_TOKEN, AppStorage } from './app-storage.service';
import { EmptyLogger, KalturaLoggerInjectionToken } from './kaltura-logger';
var KalturaCommonModule = /** @class */ (function () {
    function KalturaCommonModule() {
    }
    KalturaCommonModule_1 = KalturaCommonModule;
    // constructor(@Optional() @SkipSelf() module : KalturaCoreModule, private appBootstrap : AppBootstrap)
    // {
    //     if (module) {
    //         throw new Error("KMCngCoreModule module imported twice.");
    //     }
    // }
    KalturaCommonModule.forRoot = function () {
        return {
            ngModule: KalturaCommonModule_1,
            providers: [
                { provide: APP_STORAGE_TOKEN, useClass: AppStorage },
                KalturaUtils,
                { provide: KalturaLoggerInjectionToken, useClass: EmptyLogger }
            ]
        };
    };
    var KalturaCommonModule_1;
    KalturaCommonModule = KalturaCommonModule_1 = __decorate([
        NgModule({
            imports: [
                CommonModule,
            ],
            declarations: [],
            exports: [],
            providers: []
        })
    ], KalturaCommonModule);
    return KalturaCommonModule;
}());
export { KalturaCommonModule };
