// @kaltura-ng/kaltura-ui/esm2015/lib/input-range/input-range.component.js
var InputRangeComponent_1;
import { __decorate, __metadata } from "tslib";
import { Component, Input } from "@angular/core";
import { NG_VALUE_ACCESSOR } from "@angular/forms";
let InputRangeComponent = InputRangeComponent_1 = class InputRangeComponent {
    constructor() {
        this.pointOvered = [];
        this.onChange = (value) => { };
        this.onTouched = () => { };
    }
    ngOnInit() {
        this.selectedOptionIndex = this.options.findIndex(option => option.isSelected);
        this.selectedOptionIndex = this.selectedOptionIndex == -1 ? 0 : this.selectedOptionIndex;
        this.selectedText = this.options[this.selectedOptionIndex].label;
    }
    getProgressBarStyle() {
        return 100 / (this.options.length - 1) * (this.selectedOptionIndex >= 0 ? this.selectedOptionIndex : 0);
    }
    pointSelected(index) {
        if (this.selectedOptionIndex > 0) {
            this.options[this.selectedOptionIndex].isSelected = false;
        }
        this.options[index].isSelected = true;
        this.selectedText = this.options[index].label;
        this.selectedOptionIndex = index;
        this.onTouched();
        this.onChange(this.options[index].value);
    }
    //From ControlValueAccessor interface
    writeValue(value) {
        const valueOptionIndex = this.options.findIndex(option => option.value == value);
        if (valueOptionIndex !== -1) {
            this.pointSelected(valueOptionIndex);
        }
    }
    //From ControlValueAccessor interface
    registerOnChange(fn) {
        this.onChange = fn;
    }
    //From ControlValueAccessor interface
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
};
__decorate([
    Input(),
    __metadata("design:type", Array)
], InputRangeComponent.prototype, "options", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], InputRangeComponent.prototype, "width", void 0);
InputRangeComponent = InputRangeComponent_1 = __decorate([
    Component({
        selector: 'kInputRange',
        template: "<div class=\"kRangeContainer\">\n  <div [style.width]=\"width\">\n    <div class=\"kRangeInput kRangeInputBlock\">\n      <div class=\"kBbar\">\n        <div class=\"kBarProgress\"\n             [style.width]=\"getProgressBarStyle() + '%'\"></div>\n        <div class=\"kBarPoints\">\n          <div class=\"kBarPointCol\" (mouseout)=\"pointOvered[i]=false\" (mouseover)=\"pointOvered[i]=true\" [style.width]=\"100 / (options.length-1) + '%'\"\n               *ngFor=\"let option of options; index as i\">\n            <div class=\"kPoint\"\n                 [ngClass]=\"{\n                 'kSelected': (pointOvered.length > 0 && pointOvered[i]) || i == selectedOptionIndex,\n                 'kBgPrimary':  i <= selectedOptionIndex\n               }\"></div>\n            <div class=\"kPointClickArea\"\n                 (click) = \"pointSelected(i)\"\n                 [kTooltip] = \"option.tooltip\"></div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n  <span class=\"kSelectedLabel\">{{selectedText}}</span>\n</div>\n",
        providers: [
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: InputRangeComponent_1,
                multi: true
            }
        ],
        styles: [".kRangeContainer{display:flex}.kRangeContainer .kRangeInput{display:inline-block;padding-top:10px;padding-bottom:50px;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.kRangeContainer .kRangeInput.kRangeInputBlock{display:block;width:100%}.kRangeContainer .kRangeInput .kBbar{height:4px;width:100%;background-color:#ccc;position:relative}.kRangeContainer .kRangeInput .kBbar .kBarProgress{background-color:#00a784;height:100%}.kRangeContainer .kRangeInput .kBbar .kBarPoints{position:absolute;left:0;top:-8px;width:100%}.kRangeContainer .kRangeInput .kBbar .kBarPoints .kBarPointCol{float:left;position:relative;height:20px}.kRangeContainer .kRangeInput .kBbar .kBarPoints .kBarPointCol:last-child{width:0%!important}.kRangeContainer .kRangeInput .kBbar .kBarPoints .kBarPointCol .kPointClickArea{position:absolute;width:45px;margin-top:-10px;height:40px;transform:translateX(-50%);cursor:pointer}.kRangeContainer .kRangeInput .kBbar .kBarPoints .kBarPointCol .kPoint{width:9px;height:9px;margin-top:1px;border-radius:50%;background-color:#ccc;position:absolute;transform:translateX(-50%);cursor:pointer;border:4px solid #fff}.kRangeContainer .kRangeInput .kBbar .kBarPoints .kBarPointCol .kPoint.kSelected{width:16px;height:16px;margin-top:-3px;border-radius:50%}.kRangeContainer .kRangeInput .kBbar .kBarPoints .kBarPointCol .kPoint.kBgPrimary{background-color:#00a784}.kRangeContainer .kRangeInput .kBbar .kBarPoints .kBarPointCol .kPoint.kBgPrimary.kSelected{border:none!important;margin-top:2px!important}.kRangeContainer .kRangeInput .kBbar .kBarPoints .kBarPointCol span{position:absolute;transform:translateX(-50%);bottom:-36px;white-space:nowrap}.kRangeContainer .kSelectedLabel{padding-left:30px;padding-top:3px;font-size:15px}"]
    }),
    __metadata("design:paramtypes", [])
], InputRangeComponent);
export { InputRangeComponent };
