// @kaltura-ng/kaltura-ui/esm5/lib/input-autofocus.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, Renderer2, Input, AfterViewInit } from '@angular/core';
var AutofocusDirective = /** @class */ (function () {
    function AutofocusDirective(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this._autofocus = true;
    }
    AutofocusDirective.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (this._autofocus) {
            setTimeout(function () {
                _this.renderer.selectRootElement(_this.el.nativeElement).focus();
            });
        }
    };
    Object.defineProperty(AutofocusDirective.prototype, "inputAutofocus", {
        set: function (allowed) {
            this._autofocus = allowed;
        },
        enumerable: true,
        configurable: true
    });
    AutofocusDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], AutofocusDirective.prototype, "inputAutofocus", null);
    AutofocusDirective = __decorate([
        Directive({
            selector: '[kAutofocus]'
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], AutofocusDirective);
    return AutofocusDirective;
}());
export { AutofocusDirective };
