// @kaltura-ng/kaltura-common/esm5/lib/rxjs/operators/cancel-on-destroy.js
import { __extends } from "tslib";
import { merge, Subject, Subscriber } from "rxjs";
var __ngOnDestroySource__ = Symbol("__ngOnDestroySource__");
var __ngOnDestroy__ = Symbol("__ngOnDestroy__");
export function cancelOnDestroy(instance, manualDestroy) {
    return function (source) { return source.lift(new CancelOnDestroyOperator(instance, manualDestroy)); };
}
var CancelOnDestroyOperator = /** @class */ (function () {
    function CancelOnDestroyOperator(instance, manualDestroy) {
        this.instance = instance;
        this.manualDestroy = manualDestroy;
        if (instance.ngOnDestroy) {
            if (!instance[__ngOnDestroySource__]) {
                instance[__ngOnDestroySource__] = new Subject();
                instance[__ngOnDestroy__] = instance.ngOnDestroy;
                instance.ngOnDestroy = function () {
                    this[__ngOnDestroy__].apply(this, arguments);
                    this[__ngOnDestroySource__].next();
                    this[__ngOnDestroySource__].complete();
                };
            }
        }
    }
    CancelOnDestroyOperator.prototype.call = function (subscriber, source) {
        return source.subscribe(new CancelOnDestroySubscriber(subscriber, this.instance, this.manualDestroy));
    };
    return CancelOnDestroyOperator;
}());
var CancelOnDestroySubscriber = /** @class */ (function (_super) {
    __extends(CancelOnDestroySubscriber, _super);
    function CancelOnDestroySubscriber(destination, _instance, manualDestroy) {
        var _this = _super.call(this, destination) || this;
        _this._instance = _instance;
        _this.manualDestroy = manualDestroy;
        var sources = manualDestroy
            ? merge(manualDestroy, _instance[__ngOnDestroySource__])
            : _instance[__ngOnDestroySource__].asObservable();
        _this.add(sources.subscribe(function () {
            destination.unsubscribe();
        }));
        return _this;
    }
    return CancelOnDestroySubscriber;
}(Subscriber));
