// @kaltura-ng/kaltura-ui/esm2015/lib/dynamic-form/dynamic-form-item.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, TemplateRef, ContentChildren, QueryList } from '@angular/core';
import { FormArray, FormGroup, Validators, FormBuilder, AbstractControl } from '@angular/forms';
import { DynamicFormService } from './dynamic-form.service';
import { DynamicFormControlBase } from './dynamic-form-control-base';
let DynamicFormItem = class DynamicFormItem {
    constructor(_formBuilder, _dynamicFormService) {
        this._formBuilder = _formBuilder;
        this._dynamicFormService = _dynamicFormService;
    }
    asDynamicSection(control) {
        return control;
    }
    asFormGroup(control) {
        return control;
    }
    ngOnInit() {
        this.isRequired = false;
        const validators = this.control.validators;
        if (this.control.validators) {
            const RequiredValidator = validators.find(validator => validator == Validators.required);
            if (RequiredValidator) {
                this.isRequired = true;
            }
        }
    }
    initItem(dynamicControl) {
        return this._dynamicFormService.toFormGroup([dynamicControl], { ignoreMultiple: true });
    }
    addItem(dynamicControl) {
        const control = this.form.controls[dynamicControl.key];
        control.push(this.initItem(dynamicControl));
        control.markAsDirty();
    }
    removeItem(dynamicControl, i) {
        const control = this.form.controls[dynamicControl.key];
        control.removeAt(i);
        control.markAsDirty();
    }
};
DynamicFormItem.ctorParameters = () => [
    { type: FormBuilder },
    { type: DynamicFormService }
];
__decorate([
    Input(),
    __metadata("design:type", DynamicFormControlBase)
], DynamicFormItem.prototype, "control", void 0);
__decorate([
    Input(),
    __metadata("design:type", FormGroup)
], DynamicFormItem.prototype, "form", void 0);
__decorate([
    ContentChildren(TemplateRef, { descendants: false }),
    __metadata("design:type", QueryList)
], DynamicFormItem.prototype, "_templates", void 0);
__decorate([
    Input(),
    __metadata("design:type", TemplateRef)
], DynamicFormItem.prototype, "labelTemplate", void 0);
DynamicFormItem = __decorate([
    Component({
        selector: 'k-dynamic-form-item',
        template: "<div *ngIf=\"control\" [formGroup]=\"form\">\n\n    <ng-container  *ngIf=\"control.controlType === 'Group' && control.allowMultiple\">\n            <div class=\"kMultiple\" [formArrayName]=\"control.key\">\n                <span class=\"kGroupTitle\" *ngIf=\"control.label\">\n                    <span [kTooltip]=\"control.description\">{{control.label}}</span>\n                </span>\n                <div class=\"kGroupContainer\"  *ngFor=\"let groupForm of asFormGroup(form.controls[control.key]).controls; let i=index\">\n\n                    <div [formGroupName]=\"i\">\n                        <k-dynamic-form-item *ngFor=\"let childControl of asDynamicSection(control).children\"\n                                          [control]=\"childControl\"\n                                          [form]=\"groupForm\">\n                            <ng-template let-item=\"item\">\n                                <ng-container *ngTemplateOutlet=\"_templates.first;context:{ item : item }\">\n                                </ng-container>\n                            </ng-template>\n                        </k-dynamic-form-item>\n                    </div>\n\n                    <div class=\"kDeleteGroup\">\n                        <a (click)=\"!form.disabled && removeItem(control,i)\" [class.disabled]=\"form.disabled\">Delete</a>\n                    </div>\n                </div>\n\n                <div class=\"kGroupButtons\">\n                    <button (click)=\"addItem(control)\" class=\"kButton\" [disabled]=\"form.disabled\">Add</button>\n                </div>\n            </div>\n    </ng-container>\n\n    <div class=\"kMultiple\" *ngIf=\"control.controlType === 'Group' && !control.allowMultiple\">\n        <span class=\"kGroupTitle\" *ngIf=\"control.label\">\n            <span [kTooltip]=\"control.description\">{{control.label}}</span>\n        </span>\n        <k-dynamic-form-item *ngFor=\"let childControl of asDynamicSection(control).children\" [control]=\"childControl\" [form]=\"form\" [labelTemplate]=\"labelTemplate\">\n            <ng-template let-item=\"item\">\n                <ng-container *ngTemplateOutlet=\"_templates.first;context:{ item : item }\"></ng-container>\n            </ng-template>\n        </k-dynamic-form-item>\n    </div>\n\n    <div class=\"kRow\" [ngClass]=\"control.styleClass\" *ngIf=\"control.controlType !== 'Group'\">\n        <div class=\"kFormLabelContainer\">\n            <span class=\"kLabel\" *ngIf=\"control.label && !!labelTemplate == false\">{{control.label}}\n                <kInputHelper *ngIf=\"control.inputHelperConfig && control.inputHelperConfig.body\" [title]=\"control.inputHelperConfig.title\">\n                    <span [innerHTML]=\"control.inputHelperConfig.body\"></span>\n                </kInputHelper>\n            </span>\n            <ng-container *ngIf=\"!!labelTemplate\">\n                <ng-container *ngTemplateOutlet=\"labelTemplate;context:{item: { control : control, form : form, isRequired : isRequired }}\"></ng-container>\n            </ng-container>\n            <span *ngIf=\"isRequired\" class=\"kFormRequiredIndicator\">*</span>\n        </div>\n        <div class=\"kMultiple\" *ngIf=\"control.allowMultiple\" [formArrayName]=\"control.key\">\n\n            <div class=\"kSingleItem\" *ngFor=\"let itemFormGroup of asFormGroup(form.controls[control.key]).controls; let i=index\" [formGroupName]=\"i\">\n                    <ng-container *ngTemplateOutlet=\"_templates.first;context:{item: { control : control, form : itemFormGroup }}\"></ng-container>\n                <a (click)=\"!form.disabled && removeItem(control,i)\" class=\"kDelete\" [class.disabled]=\"form.disabled\">Delete</a>\n            </div>\n            <div>\n                <button class=\"kButton\" (click)=\"!form.disabled && addItem(control)\" [disabled]=\"form.disabled\">Add</button>\n            </div>\n        </div>\n\n        <div class=\"kControl\" [ngClass]=\"[control.controlType, control.key + 'Control']\" *ngIf=\"!control.allowMultiple\">\n            <ng-container *ngTemplateOutlet=\"_templates.first;context:{item: { control : control, form : form }}\"></ng-container>\n        </div>\n        <!--<div class=\"errorMessage\" *ngIf=\"form.controls[control.key].valid\">{{control.label}} is required</div>-->\n    </div>\n</div>\n\n",
        styles: [""]
    }),
    __metadata("design:paramtypes", [FormBuilder, DynamicFormService])
], DynamicFormItem);
export { DynamicFormItem };
