// @kaltura-ng/kaltura-ui/esm5/lib/sticky/sticky.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { KalturaCommonModule } from '@kaltura-ng/kaltura-common';
import { StickyComponent } from './components/sticky.component';
import { StickyDirective } from './directives/sticky.directive';
import { StickyScrollService } from './services/sticky-scroll.service';
var StickyModule = /** @class */ (function () {
    function StickyModule() {
    }
    StickyModule_1 = StickyModule;
    StickyModule.forRoot = function () {
        return {
            ngModule: StickyModule_1,
            providers: [
                StickyScrollService
            ]
        };
    };
    var StickyModule_1;
    StickyModule = StickyModule_1 = __decorate([
        NgModule({
            imports: [
                KalturaCommonModule
            ],
            declarations: [
                StickyComponent,
                StickyDirective
            ],
            exports: [
                StickyComponent,
                StickyDirective
            ],
            providers: []
        })
    ], StickyModule);
    return StickyModule;
}());
export { StickyModule };
