// @kaltura-ng/kaltura-ui/esm2015/lib/details-bar/details-bar.component.js
import { __decorate, __metadata } from "tslib";
import { Component, ElementRef, Input, QueryList, ContentChildren, ViewChild, HostListener } from '@angular/core';
import { DetailInfoComponent } from './detail-info.component';
let DetailsBarComponent = class DetailsBarComponent {
    constructor() {
        this.basicDetailsLabel = "Basic Details";
        this.moreDetailsLabel = "More Details";
        this._shouldUpdateItems = false;
        this._showMore = false;
        this._showBasic = false;
        this.lineScroll = 0;
        this.disableScroll = false;
    }
    set data(dataObj) {
        this._data = dataObj;
        this._shouldUpdateItems = true;
        this.reset();
    }
    onWindowResize() {
        this.reset();
    }
    ngAfterViewChecked() {
        if (this._shouldUpdateItems) {
            this._shouldUpdateItems = false;
            setTimeout(() => {
                this.items.forEach(item => {
                    item._setData(this._data);
                });
            }, 0);
        }
    }
    ngAfterContentInit() {
        this.updateLayout();
        this._itemsChangesSubscription = this.items.changes.subscribe(changes => {
            this._shouldUpdateItems = true;
        });
    }
    _updateItems() {
        if (this.items) {
            this.items.forEach(item => item._setData(this._data));
        }
    }
    updateLayout() {
        //we use a cancelable interval to improve performances on window resize
        if (this.showMoreCheckIntervalID) {
            clearInterval(this.showMoreCheckIntervalID);
            this.showMoreCheckIntervalID = null;
        }
        this.showMoreCheckIntervalID = setTimeout(() => {
            this.items.forEach(item => {
                item.isLastItem = false;
            });
            this.items.last.isLastItem = true;
            const marginTop = parseInt(window.getComputedStyle(this.dataWrapper.nativeElement).top);
            const elementHeight = this.dataWrapper.nativeElement.children.length ? this.dataWrapper.nativeElement.children[0].clientHeight : 0;
            this._showMore = this.dataWrapper.nativeElement.clientHeight > this.dataPanel.nativeElement.getBoundingClientRect().height && Math.abs(marginTop) < (this.dataWrapper.nativeElement.clientHeight + marginTop);
            this._showBasic = this.dataWrapper.nativeElement.clientHeight > this.dataPanel.nativeElement.getBoundingClientRect().height && marginTop < 0;
            // code to remove last separators in each line
            let topArr = [];
            if (this.dataWrapper.nativeElement.children.length) {
                for (let i = 0; i < this.dataWrapper.nativeElement.children.length; i++) {
                    const elm = this.dataWrapper.nativeElement.children[i];
                    const top = elm.getBoundingClientRect().top;
                    topArr.push(top);
                }
                for (let i = 0; i < topArr.length - 1; i++) {
                    if (topArr[i] < topArr[i + 1] && this.items.length >= i) {
                        this.items.forEach((item, index) => {
                            if (i === index) {
                                item.isLastItem = true;
                            }
                        });
                    }
                }
            }
            this.showMoreCheckIntervalID = null;
        }, 100);
    }
    show(direction) {
        if (!this.disableScroll) {
            this.disableScroll = true;
            if (direction === "more") {
                this.lineScroll++;
            }
            else {
                this.lineScroll--;
            }
            this.dataWrapper.nativeElement.style.top = this.dataWrapper.nativeElement.children[0].clientHeight * (-1) * this.lineScroll + "px";
            setTimeout(() => {
                this.updateLayout(); // allow animation to finish before recalculating
                this.disableScroll = false;
            }, 350);
        }
    }
    reset() {
        this.lineScroll = 0;
        this.dataWrapper.nativeElement.style.top = "0px";
        setTimeout(() => {
            this.updateLayout(); // allow animation to finish before recalculating
        }, 350);
    }
    ngOnDestroy() {
        if (this._itemsChangesSubscription) {
            this._itemsChangesSubscription.unsubscribe();
            this._itemsChangesSubscription = null;
        }
    }
};
__decorate([
    ContentChildren(DetailInfoComponent),
    __metadata("design:type", QueryList)
], DetailsBarComponent.prototype, "items", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], DetailsBarComponent.prototype, "basicDetailsLabel", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], DetailsBarComponent.prototype, "moreDetailsLabel", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], DetailsBarComponent.prototype, "data", null);
__decorate([
    ViewChild('dataPanel', { static: true }),
    __metadata("design:type", ElementRef)
], DetailsBarComponent.prototype, "dataPanel", void 0);
__decorate([
    ViewChild('dataWrapper', { static: true }),
    __metadata("design:type", ElementRef)
], DetailsBarComponent.prototype, "dataWrapper", void 0);
__decorate([
    HostListener("window:resize", []),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], DetailsBarComponent.prototype, "onWindowResize", null);
DetailsBarComponent = __decorate([
    Component({
        selector: 'k-details-bar',
        template: "<div class=\"kContent kDetailsBar\" #dataPanel>\n    <i class=\"kLink kIcondropdown_arrow_bottom\" [class.disabled]=\"!_showMore\" [class.hidden]=\"!_showMore && !_showBasic\" (click)=\"show('more')\"></i>\n    <i class=\"kLink kIcondropdown_arrow_top\" [class.disabled]=\"!_showBasic\" [class.hidden]=\"!_showMore && !_showBasic\" (click)=\"show('basic')\"></i>\n    <div class=\"kItemsWrapper\" #dataWrapper>\n        <ng-content *ngIf=\"_data\" ></ng-content>\n    </div>\n</div>",
        styles: [".kDetailsBar{display:block;height:26px;line-height:26px;overflow-y:hidden;font-size:14px;color:#666}.kDetailsBar .kLink{font-size:14px;float:right;margin-left:8px;border:0;cursor:pointer;margin-top:5px}.kDetailsBar .kLink.disabled{opacity:.4;pointer-events:none}.kDetailsBar .kLink.hidden{opacity:0;pointer-events:none}.kDetailsBar .kItemsWrapper{position:relative;flex:1 1 auto;height:auto;display:flex;flex-flow:wrap;transition:top .3s}"]
    }),
    __metadata("design:paramtypes", [])
], DetailsBarComponent);
export { DetailsBarComponent };
