// @kaltura-ng/kaltura-ui/esm5/lib/area-blocker/area-blocker.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, OnInit, ViewEncapsulation, ViewChild, ElementRef, Renderer2 } from '@angular/core';
import { AreaBlockerMessage } from './area-blocker-message';
var AreaBlockerComponent = /** @class */ (function () {
    function AreaBlockerComponent(_renderer) {
        this._renderer = _renderer;
        this.bodyScroll = false;
        this.spinnerMarginTop = 0;
    }
    Object.defineProperty(AreaBlockerComponent.prototype, "showLoader", {
        get: function () {
            return this._showLoader;
        },
        set: function (value) {
            var _this = this;
            // once showLoader is set to true, use a timeout so *ngIf will cause the HTML to render and then calculate area blocker width
            if (value) {
                this._showLoaderHandler = setTimeout(function () {
                    _this._showLoaderHandler = null;
                    if (_this.bodyScroll && !!_this.areaBlockerContainer) {
                        var rect = _this.areaBlockerContainer.nativeElement.getBoundingClientRect();
                        if (rect.width < document.body.clientWidth && _this.spinnerContainer) {
                            _this._renderer.setStyle(_this.spinnerContainer.nativeElement, 'left', rect.left + rect.width / 2 + "px");
                        }
                    }
                    _this._renderer.setStyle(_this.spinnerContainer.nativeElement, 'opacity', '1'); // show the spinner only after its position is calculated to prevent seeing it jumps...
                }, 0);
            }
            else {
                if (this._showLoaderHandler) {
                    clearTimeout(this._showLoaderHandler);
                    this._showLoaderHandler = null;
                }
                if (this.bodyScroll && this.spinnerContainer) {
                    this._renderer.setStyle(this.spinnerContainer.nativeElement, 'opacity', '0'); // hide the spinner so we won't see it jump next time its position is recalculated
                }
            }
            this._showLoader = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AreaBlockerComponent.prototype, "message", {
        set: function (value) {
            var _this = this;
            if (typeof value === 'string') {
                this._message = { title: 'Error', message: value, buttons: [{ label: 'Dismiss', action: function () { _this._message = null; } }] };
            }
            else if (value instanceof AreaBlockerMessage) {
                this._message = value;
            }
            else {
                this._message = null;
            }
        },
        enumerable: true,
        configurable: true
    });
    AreaBlockerComponent.prototype.handleAction = function (button) {
        if (button) {
            button.action();
        }
    };
    AreaBlockerComponent.prototype.ngOnInit = function () {
    };
    AreaBlockerComponent.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    __decorate([
        ViewChild('areaBlockerContainer', { static: false }),
        __metadata("design:type", ElementRef)
    ], AreaBlockerComponent.prototype, "areaBlockerContainer", void 0);
    __decorate([
        ViewChild('spinnerContainer', { static: false }),
        __metadata("design:type", ElementRef)
    ], AreaBlockerComponent.prototype, "spinnerContainer", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], AreaBlockerComponent.prototype, "bodyScroll", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], AreaBlockerComponent.prototype, "spinnerMarginTop", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], AreaBlockerComponent.prototype, "classes", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], AreaBlockerComponent.prototype, "showLoader", null);
    __decorate([
        Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], AreaBlockerComponent.prototype, "message", null);
    AreaBlockerComponent = __decorate([
        Component({
            selector: 'k-area-blocker',
            template: "<ng-content></ng-content>\n<div #areaBlockerContainer *ngIf=\"showLoader || _message\" class=\"kAreaBlockerContainer\" [class.kAreaBlockerContainerError]=\"_message\" [ngClass]=\"classes\">\n  <div *ngIf=\"showLoader && !_message\" class=\"loading-backdrop show\">\n    <div #spinnerContainer class='spinner-container' [class.bodyScroll]=\"bodyScroll\" [style.marginTop]=\"spinnerMarginTop+'px'\">\n      <div class='spinner'><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></div>\n    </div>\n  </div>\n  <div  class=\"kErrorMessageContainer\" *ngIf=\"_message\" [class.bodyScroll]=\"bodyScroll\">\n    <div class=\"kErrorMessage\">\n      <div class=\"kErrorMessageTitle\">{{ _message.title }}</div>\n      <div class=\"kErrorMessageText\">{{ _message.message }}</div>\n      <div class=\"kErrorButtons\">\n        <button *ngFor=\"let button of _message.buttons\" (click)=\"handleAction(button)\" [ngClass]=\"button.classes\">{{button.label}}</button>\n      </div>\n    </div>\n  </div>\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            styles: ["k-area-blocker{position:relative;width:100%;height:100%;display:flex;flex:1 1 auto}k-area-blocker>.kAreaBlockerContainer{display:flex;flex-direction:row;align-items:center;justify-content:center;width:100%;height:100%;background-color:rgba(255,255,255,.7);position:absolute;top:0;left:0;z-index:510}@-webkit-keyframes kaltura-spinner{0%{transform:rotate(0) scale(.7);opacity:1}100%,70%{transform:rotate(360deg) scale(.7);opacity:1}82%{transform:rotate(360deg) scale(0);opacity:0}87%{transform:rotate(360deg) scale(.9);opacity:1}}@keyframes kaltura-spinner{0%{transform:rotate(0) scale(.7);opacity:1}100%,70%{transform:rotate(360deg) scale(.7);opacity:1}82%{transform:rotate(360deg) scale(0);opacity:0}87%{transform:rotate(360deg) scale(.9);opacity:1}}k-area-blocker>.kAreaBlockerContainer .loading-backdrop{position:absolute;top:0;left:0;width:100%;height:100%}k-area-blocker>.kAreaBlockerContainer .loading-backdrop .spinner-container{display:block;opacity:0;transform:translate3d(-50px,-50px,0)}k-area-blocker>.kAreaBlockerContainer .loading-backdrop .spinner-container.bodyScroll{position:fixed}k-area-blocker>.kAreaBlockerContainer .spinner{width:100px;height:100px;position:relative;-webkit-animation:2.5s infinite kaltura-spinner;animation:2.5s infinite kaltura-spinner}k-area-blocker>.kAreaBlockerContainer .spinner span{width:8px;height:8px;background-color:#fff;display:block;border-radius:8px;position:absolute;top:50%;left:50%;margin-top:-4px;margin-left:-4px}k-area-blocker>.kAreaBlockerContainer .spinner span:nth-child(1){transform:rotate(45deg) translateX(-25px) translateY(-25px);background-color:#da1f26}k-area-blocker>.kAreaBlockerContainer .spinner span:nth-child(2){transform:rotate(90deg) translateX(-25px) translateY(-25px);background-color:#06a885}k-area-blocker>.kAreaBlockerContainer .spinner span:nth-child(3){transform:rotate(135deg) translateX(-25px) translateY(-25px);background-color:#009344}k-area-blocker>.kAreaBlockerContainer .spinner span:nth-child(4){transform:rotate(180deg) translateX(-25px) translateY(-25px);background-color:#f8a61a}k-area-blocker>.kAreaBlockerContainer .spinner span:nth-child(5){transform:rotate(225deg) translateX(-25px) translateY(-25px);background-color:#1b4a97}k-area-blocker>.kAreaBlockerContainer .spinner span:nth-child(6){transform:rotate(270deg) translateX(-25px) translateY(-25px);background-color:#00abcc}k-area-blocker>.kAreaBlockerContainer .spinner span:nth-child(7){transform:rotate(315deg) translateX(-25px) translateY(-25px);background-color:#b1d238}k-area-blocker>.kAreaBlockerContainer .spinner span:nth-child(8){transform:rotate(360deg) translateX(-25px) translateY(-25px);background-color:#fcd203}k-area-blocker>.kAreaBlockerContainer .kErrorMessageContainer{display:flex;align-items:center;justify-content:center;flex-direction:column;z-index:1001}k-area-blocker>.kAreaBlockerContainer .kErrorMessageContainer.bodyScroll{margin-top:0;position:fixed;top:calc(50vh - 90px);left:calc(50vw - 293px)}k-area-blocker>.kAreaBlockerContainer .kErrorMessage{width:586px;background-color:#fff;border-radius:4px;box-shadow:0 16px 20px 0 rgba(0,0,0,.3);padding:60px 32px 32px;box-sizing:border-box;text-align:center}k-area-blocker>.kAreaBlockerContainer .kErrorMessageTitle{color:#333;font-size:20px;font-weight:700;line-height:24px}k-area-blocker>.kAreaBlockerContainer .kErrorMessageText{max-height:250px;min-height:30px;padding:4px;overflow-y:auto;color:#999;font-size:15px;font-weight:700;line-height:18px;margin-top:26px;white-space:pre-wrap}k-area-blocker>.kAreaBlockerContainer .kErrorButtons{margin-top:44px}k-area-blocker>.kAreaBlockerContainer .kErrorButtons button{cursor:pointer;min-width:100px;outline:0;color:#333;font-size:14px;font-weight:700;line-height:17px;text-align:center;height:34px;border:1px solid #ccc;border-radius:3px;background-color:#fff;box-sizing:border-box;padding:0 12px;margin:0 8px}"]
        }),
        __metadata("design:paramtypes", [Renderer2])
    ], AreaBlockerComponent);
    return AreaBlockerComponent;
}());
export { AreaBlockerComponent };
