// @kaltura-ng/kaltura-ui/esm5/lib/file-dialog/file-dialog.component.js
import { __decorate, __metadata } from "tslib";
import { Output, EventEmitter, Component, ElementRef, Input, OnInit, HostListener } from '@angular/core';
var FileDialogComponent = /** @class */ (function () {
    function FileDialogComponent(elementRef) {
        this.elementRef = elementRef;
        this.filter = "";
        this.allowMultiple = false;
        this.onFileSelected = new EventEmitter();
    }
    FileDialogComponent.prototype.open = function (event) {
        if (event) {
            event.stopPropagation();
            event.preventDefault();
        }
        this.elementRef.nativeElement.firstElementChild.value = "";
        this.elementRef.nativeElement.firstElementChild.click();
    };
    FileDialogComponent.prototype._fileInputChange = function (event) {
        if (event.currentTarget.files && event.currentTarget.files.length) {
            this.onFileSelected.emit(event.currentTarget.files);
        }
    };
    FileDialogComponent.prototype.ngOnInit = function () {
    };
    FileDialogComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], FileDialogComponent.prototype, "filter", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], FileDialogComponent.prototype, "allowMultiple", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], FileDialogComponent.prototype, "onFileSelected", void 0);
    FileDialogComponent = __decorate([
        Component({
            selector: 'kFileDialog',
            template: "<input tabindex=\"-1\" [attr.multiple]=\"allowMultiple ? true : null\" type=\"file\" accept=\"{{filter}}\" (change)=\"_fileInputChange($event)\"/>\n",
            styles: ["input{position:fixed;left:-1000px;top:-1000px}"]
        }),
        __metadata("design:paramtypes", [ElementRef])
    ], FileDialogComponent);
    return FileDialogComponent;
}());
export { FileDialogComponent };
