// @kaltura-ng/kaltura-ui/esm2015/lib/file-dialog/file-dialog.component.js
import { __decorate, __metadata } from "tslib";
import { Output, EventEmitter, Component, ElementRef, Input, OnInit, HostListener } from '@angular/core';
let FileDialogComponent = class FileDialogComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.filter = "";
        this.allowMultiple = false;
        this.onFileSelected = new EventEmitter();
    }
    open(event) {
        if (event) {
            event.stopPropagation();
            event.preventDefault();
        }
        this.elementRef.nativeElement.firstElementChild.value = "";
        this.elementRef.nativeElement.firstElementChild.click();
    }
    _fileInputChange(event) {
        if (event.currentTarget.files && event.currentTarget.files.length) {
            this.onFileSelected.emit(event.currentTarget.files);
        }
    }
    ngOnInit() {
    }
};
FileDialogComponent.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Input(),
    __metadata("design:type", String)
], FileDialogComponent.prototype, "filter", void 0);
__decorate([
    Input(),
    __metadata("design:type", Boolean)
], FileDialogComponent.prototype, "allowMultiple", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], FileDialogComponent.prototype, "onFileSelected", void 0);
FileDialogComponent = __decorate([
    Component({
        selector: 'kFileDialog',
        template: "<input tabindex=\"-1\" [attr.multiple]=\"allowMultiple ? true : null\" type=\"file\" accept=\"{{filter}}\" (change)=\"_fileInputChange($event)\"/>\n",
        styles: ["input{position:fixed;left:-1000px;top:-1000px}"]
    }),
    __metadata("design:paramtypes", [ElementRef])
], FileDialogComponent);
export { FileDialogComponent };
