// @kaltura-ng/kaltura-ui/esm5/lib/popup-widget/popup-widget.component.js
import { __decorate, __metadata } from "tslib";
import { Component, AfterViewInit, EventEmitter, OnDestroy, Input, Output, ElementRef, HostListener, TemplateRef, ContentChild, Renderer2, HostBinding } from "@angular/core";
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { PopupWidgetLayout } from './popup-widget-layout';
import 'rxjs/add/observable/fromEvent';
import { cancelOnDestroy } from '@kaltura-ng/kaltura-common';
export var PopupWidgetStates = {
    "Open": "open",
    "BeforeClose": "beforeClose",
    "Close": "close",
    "Disabled": "disabled"
};
var WINDOW_GUTTER = 16;
var PopupWidgetComponent = /** @class */ (function () {
    function PopupWidgetComponent(popup, renderer) {
        var _this = this;
        this.popup = popup;
        this.renderer = renderer;
        this.transparent = false;
        this.popupHeight = 'auto';
        this.showTooltip = false;
        this.preventPageScroll = false;
        this.modal = false;
        this.slider = false;
        this.fullScreen = false;
        this.closeBtn = true;
        this.closeBtnInside = false;
        this.closeOnClickOutside = true;
        this.closeOnResize = false;
        this.closeOnBrowserNav = true;
        this.targetOffset = { 'x': 0, 'y': 0 };
        this.childrenPopups = [];
        this.trigger = 'click';
        this.placement = { x: 'right', y: 'bottom' };
        this.closeOnScroll = false;
        this.opened = false;
        this.closed = false;
        this._toggleFunc = this.toggle.bind(this);
        this._openFunc = this.open.bind(this);
        this._closeFunc = this.close.bind(this);
        this._viewInitialize = false;
        this.onOpen = new EventEmitter();
        this.onClose = new EventEmitter();
        this._stateChangeSubscription = null;
        this._statechange = new BehaviorSubject({ state: '' });
        this.state$ = this._statechange.asObservable();
        this._statechange.next({ state: PopupWidgetStates.Close });
        this.state$
            .pipe(cancelOnDestroy(this))
            .subscribe(function (_a) {
            var state = _a.state;
            _this.closed = state === PopupWidgetStates.Close;
            _this.opened = state === PopupWidgetStates.Open;
        });
    }
    Object.defineProperty(PopupWidgetComponent.prototype, "targetRef", {
        get: function () { return this._targetRef; },
        set: function (targetRef) {
            if (this.trigger === 'click') {
                if (this._targetRef) {
                    this._targetRef.removeEventListener('click', this._toggleFunc);
                }
                this._targetRef = targetRef;
                if (this._targetRef) {
                    this._targetRef.addEventListener('click', this._toggleFunc);
                }
            }
            else if (this.trigger === 'hover') {
                if (this._targetRef) {
                    this._targetRef.removeEventListener('mouseover', this._openFunc);
                    this._targetRef.removeEventListener('mouseout', this._closeFunc);
                }
                this._targetRef = targetRef;
                if (this._targetRef) {
                    this._targetRef.addEventListener('mouseover', this._openFunc);
                    this._targetRef.addEventListener('mouseout', this._closeFunc);
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopupWidgetComponent.prototype, "parentPopup", {
        get: function () { return this._parentPopup; },
        set: function (parentPopup) {
            var _this = this;
            this._parentPopup = parentPopup;
            if (this._stateChangeSubscription) {
                this._stateChangeSubscription.unsubscribe();
                this._stateChangeSubscription = null;
            }
            if (this._parentPopup) {
                this._stateChangeSubscription = this._parentPopup.state$.subscribe(function (event) {
                    if (event.state === PopupWidgetStates.Close) {
                        _this.close();
                    }
                });
            }
        },
        enumerable: true,
        configurable: true
    });
    // public API methods
    PopupWidgetComponent.prototype.open = function () {
        var _this = this;
        if (this.isEnabled && this.validate()) {
            // handle auto height
            if (!this.popupHeight || this.popupHeight === 'auto') {
                if (this.slider) {
                    this._popupWidgetHeight = 'calc(100vh - 80px)';
                }
                else {
                    this._popupWidgetHeight = 'auto';
                }
            }
            else {
                this._popupWidgetHeight = this.popupHeight + "px";
            }
            this._saveScrollPosition = window.pageYOffset || document.documentElement.scrollTop;
            // set location according to targetRef
            var parentLeft = this.appendTo && !this.modal ? this.appendTo.getBoundingClientRect().left : 0;
            var parentTop = this.appendTo && !this.modal ? this.appendTo.getBoundingClientRect().top : 0;
            // center on screen if no targetRef was defined
            if (!this._targetRef) {
                if (this.fullScreen) {
                    this.modal = false;
                    this.preventPageScroll = true;
                    this.renderer.addClass(this.popup.nativeElement, 'fullScreen');
                }
                else {
                    this.popup.nativeElement.style.marginLeft = window.innerWidth / 2 - this.popupWidth / 2 + 'px';
                    if (this.slider) {
                        window.scrollTo(0, 0);
                        this.popup.nativeElement.style.top = "auto";
                        this.closeBtn = false;
                        this.preventPageScroll = true;
                        this.popup.nativeElement.style.bottom = this.popupHeight !== 'auto' ? this.popupHeight * -1 + "px" : "-1000px";
                        setTimeout(function () {
                            _this.popup.nativeElement.style.bottom = 0 + "px"; // use timeout to invoke animation
                        }, 0);
                    }
                    else {
                        var marginTop = (window.innerHeight / 2 - 100 / 2);
                        this.popup.nativeElement.style.marginTop = marginTop + 'px';
                        this.popup.nativeElement.style.position = "fixed";
                    }
                }
            }
            else {
                this.popup.nativeElement.style.marginLeft = this._targetRef.getBoundingClientRect().left - parentLeft + this.targetOffset['x'] + 'px';
                this.popup.nativeElement.style.marginTop = this._targetRef.getBoundingClientRect().top - parentTop + this.targetOffset['y'] + 'px';
                this.popup.nativeElement.style.position = "absolute";
            }
            this.popup.nativeElement.style.zIndex = PopupWidgetLayout.getPopupZindex(this.fullScreen);
            // handle modal
            if (!this._modalOverlay) {
                if (this.trigger !== 'hover') {
                    this._modalOverlay = document.createElement('div');
                    if (this.modal || this.slider) {
                        this._modalOverlay.className = "kPopupWidgetModalOverlay";
                    }
                    else {
                        this._modalOverlay.className = "kPopupWidgetModalOverlay kTransparent";
                    }
                    this._modalOverlay.style.zIndex = this.popup.nativeElement.style.zIndex - 1;
                    if (!this.slider && this.closeOnClickOutside) {
                        this._modalOverlay.addEventListener("mousedown", function (event) {
                            event.stopPropagation();
                            _this.close();
                        });
                    }
                    if (this.appendTo) {
                        this.appendChild(this._modalOverlay, this.appendTo);
                    }
                    else {
                        document.body.appendChild(this._modalOverlay);
                    }
                    if (this.modal || this.slider) {
                        PopupWidgetLayout.increaseModalCount();
                    }
                }
            }
            // prevent page scroll
            if (this.preventPageScroll) {
                document.body.style.overflowY = 'hidden';
                document.body.style.position = 'fixed';
            }
            setTimeout(function () {
                _this.addClickOutsideSupport();
            }, 0);
            this.onOpen.emit(); // dispatch onOpen event (API)
            this._statechange.next({ state: PopupWidgetStates.Open });
            if (this.closeOnScroll && this.isShow) {
                window.addEventListener('scroll', this._closeFunc);
            }
        }
        if (!this.slider && !this.fullScreen && this.popup.nativeElement) {
            this.popup.nativeElement.style.opacity = 0;
        }
        // auto positioning need first the dom to render
        setTimeout(function () {
            _this.setPosition();
        }, 0);
    };
    PopupWidgetComponent.prototype.close = function (context, reason) {
        var _this = this;
        if (context === void 0) { context = null; }
        if (reason === void 0) { reason = null; }
        if (this.isEnabled && this.isShow) {
            window.removeEventListener('scroll', this._closeFunc);
            if (this.fullScreen) {
                this.renderer.removeClass(this.popup.nativeElement, 'fullScreen');
            }
            // allow cancelling the close operation
            var beforeCloseContext = { "allowClose": true };
            this._statechange.next({ state: PopupWidgetStates.BeforeClose, context: beforeCloseContext });
            if (beforeCloseContext['allowClose']) {
                // close children popups if exist
                if (this.childrenPopups.length) {
                    this.childrenPopups.forEach(function (popup) {
                        popup.close();
                    });
                }
                this.removeClickOutsideSupport();
                this.restorePageScroll(true);
                this.onClose.emit(); // dispatch onClose event (API)
                var timeout = 0;
                if (this.slider) {
                    this.popup.nativeElement.style.bottom = this.popupHeight !== 'auto' ? this.popupHeight * -1 + "px" : "-1000px";
                    timeout = 300;
                }
                if (this.modal && !this.slider) {
                    PopupWidgetLayout.decreaseModalCount();
                }
                setTimeout(function () {
                    // remove modal
                    if (_this._modalOverlay) {
                        if (_this.appendTo) {
                            _this.removeChild(_this._modalOverlay, _this.appendTo);
                        }
                        else {
                            document.body.removeChild(_this._modalOverlay);
                        }
                        _this._modalOverlay = null;
                    }
                    if (_this.slider) {
                        PopupWidgetLayout.decreaseModalCount();
                    }
                    _this._statechange.next({ state: PopupWidgetStates.Close, context: context, reason: reason }); // use timeout to prevent valueChangeAfterChecked error
                }, timeout);
            }
        }
    };
    PopupWidgetComponent.prototype.toggle = function () {
        if (this.isEnabled) {
            this.isShow ? this.close() : this.open();
        }
    };
    Object.defineProperty(PopupWidgetComponent.prototype, "isShow", {
        get: function () {
            return this._statechange.getValue().state !== PopupWidgetStates.Close;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PopupWidgetComponent.prototype, "isEnabled", {
        get: function () {
            return this._statechange.getValue().state !== PopupWidgetStates.Disabled;
        },
        enumerable: true,
        configurable: true
    });
    PopupWidgetComponent.prototype.onWindowResize = function () {
        if (this.closeOnResize) {
            this.close();
        }
        else {
            this.setPosition();
        }
    };
    PopupWidgetComponent.prototype.onPopState = function (event) {
        if (this.closeOnBrowserNav) {
            this.close();
        }
    };
    // component lifecycle events
    PopupWidgetComponent.prototype.ngAfterViewInit = function () {
        this._viewInitialize = true;
        if (this.validate()) {
            if (this.appendTo && !this.modal) {
                this.appendChild(this.popup.nativeElement, this.appendTo);
            }
            else {
                document.body.appendChild(this.popup.nativeElement);
                if (this.appendTo) {
                    console.warn("[kaltura] -> Ignoring append to " + this.appendTo + " since popup is set to modal=true."); // keep warning
                }
            }
            if (this.fullScreen) {
                this.popup.nativeElement.style.width = '0px';
                this.popup.nativeElement.style.height = '0px';
            }
        }
    };
    PopupWidgetComponent.prototype.ngOnDestroy = function () {
        if (this._targetRef) {
            this._targetRef.removeEventListener('click', this._toggleFunc);
            this._targetRef.removeEventListener('mouseover', this._openFunc);
            this._targetRef.removeEventListener('mouseout', this._closeFunc);
        }
        if (this._stateChangeSubscription) {
            this._stateChangeSubscription.unsubscribe();
        }
        this._statechange.complete();
        this.removeClickOutsideSupport();
        this.restorePageScroll(false);
        if (this.modal && this._modalOverlay) {
            document.body.removeChild(this._modalOverlay);
            this._modalOverlay = null;
        }
        if (this.appendTo && !this.modal) {
            this.removeChild(this.popup.nativeElement, this.appendTo);
        }
        else {
            if (this._viewInitialize) {
                document.body.removeChild(this.popup.nativeElement);
            }
        }
        window.removeEventListener('scroll', this._closeFunc);
    };
    // private methods
    PopupWidgetComponent.prototype.addClickOutsideSupport = function () {
        if (this.closeOnClickOutside) {
            this.popup.nativeElement.addEventListener('mousedown', this.blockMouseDownHandler);
            if (this.targetRef) {
                this.targetRef.addEventListener('mousedown', this.blockMouseDownHandler);
            }
        }
    };
    PopupWidgetComponent.prototype.blockMouseDownHandler = function (event) {
        event.stopPropagation();
    };
    PopupWidgetComponent.prototype.removeClickOutsideSupport = function () {
        if (this.closeOnClickOutside) {
            this.popup.nativeElement.removeEventListener('mousedown', this.blockMouseDownHandler);
            if (this.targetRef) {
                this.targetRef.removeEventListener('mousedown', this.blockMouseDownHandler);
            }
        }
    };
    PopupWidgetComponent.prototype.restorePageScroll = function (restoreScrollPosition) {
        if (restoreScrollPosition === void 0) { restoreScrollPosition = true; }
        if (this.preventPageScroll) {
            document.body.style.overflowY = 'auto';
            document.body.style.position = '';
            if (restoreScrollPosition) {
                window.scrollTo(0, this._saveScrollPosition);
            }
        }
    };
    PopupWidgetComponent.prototype.validate = function () {
        var valid = typeof this.popupWidth !== 'undefined' || this.fullScreen;
        if (!valid) {
            this._statechange.next({ state: PopupWidgetStates.Disabled });
            throw "Popup widget error: missing required parameters. Verify popupWidth is defined.";
        }
        return valid;
    };
    PopupWidgetComponent.prototype.appendChild = function (element, target) {
        if (this.isElement(target)) {
            target.appendChild(element);
        }
        else {
            throw 'Cannot append ' + target + ' to ' + element;
        }
    };
    PopupWidgetComponent.prototype.removeChild = function (element, target) {
        if (this.isElement(target)) {
            target.removeChild(element);
        }
        else {
            throw 'Cannot remove ' + target + ' from ' + element;
        }
    };
    PopupWidgetComponent.prototype.isElement = function (obj) {
        return (typeof HTMLElement === "object" ? obj instanceof HTMLElement :
            obj && typeof obj === "object" && obj !== null && obj.nodeType === 1 && typeof obj.nodeName === "string");
    };
    PopupWidgetComponent.prototype.setPosition = function () {
        var popupHeight = this.popup.nativeElement.clientHeight;
        var popupWidth = this.popup.nativeElement.clientWidth;
        if (this.modal || this.slider) {
            this.popup.nativeElement.style.marginLeft = window.innerWidth / 2 - this.popupWidth / 2 + 'px';
            if (this.modal) {
                var modalMarginTop = (window.innerHeight / 2 -
                    popupHeight / 2);
                this.popup.nativeElement.style.marginTop = Math.round(modalMarginTop) + 'px';
                this.popup.nativeElement.style.opacity = 1;
            }
            return;
        }
        if (!this._targetRef)
            return;
        var popupBox = this.popup.nativeElement.getBoundingClientRect();
        var targetRefBox = this._targetRef.getBoundingClientRect();
        var parentTop = this.appendTo && !this.modal ? this.appendTo.getBoundingClientRect().top : 0;
        var parentLeft = this.appendTo && !this.modal ? this.appendTo.getBoundingClientRect().left : 0;
        var popupMarginTop;
        var popupMarginLeft;
        var popupLeftMargins = {};
        var popupTopMargins = {};
        popupTopMargins.top = (targetRefBox.top
            - parentTop
            - popupHeight
            + this.targetRef.offsetHeight
            - this.targetOffset['y']);
        popupTopMargins.bottom = (targetRefBox.top
            - parentTop
            + this.targetOffset['y']);
        popupTopMargins.center = (targetRefBox.top
            - parentTop
            - popupHeight / 2
            + this.targetRef.offsetHeight / 2
            + this.targetOffset['y']);
        popupLeftMargins.left = (targetRefBox.left
            - parentLeft
            - popupWidth
            + this.targetRef.clientWidth
            - this.targetOffset['x']);
        popupLeftMargins.right = (targetRefBox.left
            - parentLeft
            + this.targetOffset['x']);
        popupLeftMargins.center = (targetRefBox.left
            - parentLeft
            - popupWidth / 2
            + this.targetOffset['x']);
        popupMarginTop = this.placement.y ? popupTopMargins[this.placement.y] : popupTopMargins.bottom;
        popupMarginLeft = this.placement.x ? popupLeftMargins[this.placement.y] : popupLeftMargins.right;
        this.popup.nativeElement.style.marginTop = Math.round(popupMarginTop) + 'px';
        this.popup.nativeElement.style.marginLeft = Math.round(popupMarginLeft) + 'px';
        this.validatePosition(popupLeftMargins, popupTopMargins);
    };
    // validate popup widget is not outside of the viewport (+ gutter) and if it is, reposition it
    PopupWidgetComponent.prototype.validatePosition = function (popupLeftMargins, popupTopMargins) {
        var popupHeight = this.popup.nativeElement.clientHeight;
        var popupWidth = this.popup.nativeElement.clientWidth;
        var popupBox = this.popup.nativeElement.getBoundingClientRect();
        var clientHeight = window.innerHeight;
        var clientWidth = window.innerWidth;
        var popupMarginTop;
        var popupMarginLeft;
        if (this.placement.y === 'top') {
            if (popupBox.top < WINDOW_GUTTER) {
                popupMarginTop = popupTopMargins.bottom;
            }
        }
        else if (this.placement.y === 'bottom') {
            if (popupBox.top + popupHeight > clientHeight - WINDOW_GUTTER && popupBox.top > clientHeight) {
                popupMarginTop = popupTopMargins.top;
            }
            else {
                popupMarginTop = popupTopMargins.bottom;
            }
        }
        else if (this.placement.y === 'center') {
            if (popupBox.top < WINDOW_GUTTER) {
                popupMarginTop = popupTopMargins.bottom;
            }
            else if (popupBox.top + popupHeight > clientHeight - WINDOW_GUTTER) {
                popupMarginTop = popupTopMargins.top;
            }
        }
        if (popupBox.left + popupWidth > clientWidth + WINDOW_GUTTER) {
            popupMarginLeft = popupLeftMargins.left;
        }
        else if (popupBox.left < WINDOW_GUTTER) {
            popupMarginLeft = popupLeftMargins.right;
        }
        this.popup.nativeElement.style.marginTop = Math.round(popupMarginTop) + 'px';
        this.popup.nativeElement.style.marginLeft = Math.round(popupMarginLeft) + 'px';
        this.popup.nativeElement.style.opacity = 1;
    };
    PopupWidgetComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], PopupWidgetComponent.prototype, "transparent", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], PopupWidgetComponent.prototype, "appendTo", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], PopupWidgetComponent.prototype, "popupWidth", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], PopupWidgetComponent.prototype, "popupHeight", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "showTooltip", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "preventPageScroll", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "modal", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "slider", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "fullScreen", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "closeBtn", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "closeBtnInside", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "closeOnClickOutside", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "closeOnResize", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "closeOnBrowserNav", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], PopupWidgetComponent.prototype, "targetOffset", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], PopupWidgetComponent.prototype, "childrenPopups", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], PopupWidgetComponent.prototype, "trigger", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], PopupWidgetComponent.prototype, "placement", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PopupWidgetComponent.prototype, "closeOnScroll", void 0);
    __decorate([
        ContentChild(TemplateRef, { static: true }),
        __metadata("design:type", TemplateRef)
    ], PopupWidgetComponent.prototype, "_template", void 0);
    __decorate([
        HostBinding('class.opened'),
        __metadata("design:type", Object)
    ], PopupWidgetComponent.prototype, "opened", void 0);
    __decorate([
        HostBinding('class.closed'),
        __metadata("design:type", Object)
    ], PopupWidgetComponent.prototype, "closed", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PopupWidgetComponent.prototype, "targetRef", null);
    __decorate([
        Input(),
        __metadata("design:type", PopupWidgetComponent),
        __metadata("design:paramtypes", [PopupWidgetComponent])
    ], PopupWidgetComponent.prototype, "parentPopup", null);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], PopupWidgetComponent.prototype, "onOpen", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], PopupWidgetComponent.prototype, "onClose", void 0);
    __decorate([
        HostListener("window:resize", []),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], PopupWidgetComponent.prototype, "onWindowResize", null);
    __decorate([
        HostListener('window:popstate', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], PopupWidgetComponent.prototype, "onPopState", null);
    PopupWidgetComponent = __decorate([
        Component({
            selector: 'kPopupWidget',
            template: "<div #container\n     class=\"kPopupWidget\"\n     [class.transparent]=\"transparent\"\n     [class.slider]=\"slider\"\n     [style.display]=\"(state$ | async)?.state !== 'close' ? 'block' : 'none'\"\n     [style.width]=\"popupWidth+'px'\"\n     [style.height]=\"_popupWidgetHeight\">\n    <div class=\"tooltipUp\" *ngIf=\"showTooltip\" [style.marginLeft]=\"popupWidth/2-12+'px'\"></div>\n    <i *ngIf=\"closeBtn\" class=\"kCloseBtn kIconclose_small\" [class.closeBtnInside]=\"closeBtnInside\" (click)=\"close()\"></i>\n    <div class=\"kPopupWidgetContent\" *ngIf=\"isShow\">\n        <ng-content *ngIf=\"!_template\"></ng-content>\n\t\t<ng-template *ngIf=\"_template\"\n\t\t\t[ngTemplateOutlet]=\"_template\">\n\t\t</ng-template>\n    </div>\n</div>\n",
            styles: [":host{position:absolute;left:0;top:0;transition:bottom .3s ease-in-out}:host.fullScreen{width:100%!important;height:100%!important;position:fixed}.kPopupWidget{background-color:#fff;border-radius:4px;box-shadow:0 16px 20px 0 rgba(0,0,0,.3)}.kPopupWidget.slider{border-bottom-right-radius:0;border-bottom-left-radius:0}.kPopupWidget.transparent{box-shadow:none;background:0}.kPopupWidget .kCloseBtn:not(.closeBtnInside){position:absolute;right:-118px;top:-62px;width:38px;height:38px;background-color:rgba(255,255,255,.2);border-radius:50%;text-align:center;padding-top:11px;padding-left:2px;box-sizing:border-box;color:#fff;cursor:pointer}.kPopupWidget .kCloseBtn:not(.closeBtnInside):hover{background-color:rgba(255,255,255,.4)}.kPopupWidget .kCloseBtn.closeBtnInside{font-size:14px;float:right;color:#999;margin-top:14px;margin-right:14px;cursor:pointer}.kPopupWidget .kCloseBtn.closeBtnInside:hover{color:#00a784}.kPopupWidget .tooltipUp{background-image:url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAOCAYAAAA1+Nx+AAAAAXNSR0IArs4c6QAAAMpJREFUOBGtk0EOwiAQRcF0Z2JXspFTcAE9erkACXfADa50b3DGMAYjlGktSUNDhvfmlyLEguG9V86544ItQnKLEZxSslCfpJQXY8yNs3fHKcrwSSl1gGcE0cRN0k1AnSNca33ChkII1xjjA5Kce0lmE5SdExwF+I5CTpJmggI+lnAU0MhJ7nNnUhVw4FzJj2AJnCP5EmS4LQ+UIL25dfAfwT9wktckb8EW8JZEbgmvSQa8/mu+OcFqM/3WcBntAPA9LdSK165l5vMFmCyyP+zXS30AAAAASUVORK5CYII=);position:absolute;margin-top:-12px;width:24px;height:12px}"]
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], PopupWidgetComponent);
    return PopupWidgetComponent;
}());
export { PopupWidgetComponent };
