// @kaltura-ng/kaltura-ui/esm5/lib/player/player.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
var KalturaPlayerComponent = /** @class */ (function () {
    function KalturaPlayerComponent() {
        this.width = 480;
        this.height = 360;
        this.cdnUrl = 'http://cdnapi.kaltura.com';
        this.flashvars = {};
        this.lazy = false;
        this.id = "";
        this.kalturaPlayerReady = new EventEmitter();
    }
    KalturaPlayerComponent.prototype.ngAfterViewInit = function () {
        if (!this.lazy) {
            this.Embed();
        }
    };
    KalturaPlayerComponent.prototype.Embed = function () {
        var _this = this;
        // validation
        if (!this.pid || !this.uiconfid || !this.entryid) {
            console.warn("Kaltura Player::Missing parameters. Please provide pid, uiconfid and entryid.");
        }
        else {
            // load player lib if doesn't exist
            if (document.getElementById("kalturaPlayerLib") === null) {
                var s = document.createElement('script');
                s.src = this.cdnUrl + "/p/" + this.pid + "/sp/" + this.pid + "00/embedIframeJs/uiconf_id/" + this.uiconfid + "/partner_id/" + this.pid;
                s.id = "kalturaPlayerLib";
                s.async = false;
                document.head.appendChild(s);
            }
            // wait for lib to load if not loaded and then embed player
            if (!this.kdp) {
                var intervalID_1 = setInterval(function () {
                    if (typeof window['kWidget'] !== "undefined") {
                        clearInterval(intervalID_1);
                        _this.doEmbed();
                    }
                }, 50);
            }
            else {
                this.doEmbed();
            }
        }
    };
    KalturaPlayerComponent.prototype.doEmbed = function () {
        var _this = this;
        window['kWidget'].embed({
            "targetId": "kaltura_player_" + this.id,
            "wid": "_" + this.pid,
            "uiconf_id": this.uiconfid,
            "flashvars": this.flashvars,
            "cache_st": Math.random(),
            "entry_id": this.entryid,
            "readyCallback": function (playerID) {
                _this.kdp = document.getElementById(playerID);
                _this.kalturaPlayerReady.emit(_this.kdp);
            }
        });
    };
    KalturaPlayerComponent.prototype.ngOnDestroy = function () {
        if (this.kdp) {
            window['kWidget'].destroy(this.kdp);
        }
    };
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], KalturaPlayerComponent.prototype, "width", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], KalturaPlayerComponent.prototype, "height", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], KalturaPlayerComponent.prototype, "pid", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], KalturaPlayerComponent.prototype, "uiconfid", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], KalturaPlayerComponent.prototype, "entryid", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], KalturaPlayerComponent.prototype, "cdnUrl", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], KalturaPlayerComponent.prototype, "flashvars", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], KalturaPlayerComponent.prototype, "lazy", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], KalturaPlayerComponent.prototype, "id", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], KalturaPlayerComponent.prototype, "kalturaPlayerReady", void 0);
    KalturaPlayerComponent = __decorate([
        Component({
            selector: 'k-player',
            template: "<div id=\"kaltura_player_{{id}}\" [style.width]=\"width+'px'\" [style.height]=\"height+'px'\" class=\"kPlayer\"></div>",
            styles: [".kPlayer{background-color:#000}"]
        }),
        __metadata("design:paramtypes", [])
    ], KalturaPlayerComponent);
    return KalturaPlayerComponent;
}());
export { KalturaPlayerComponent };
