// @kaltura-ng/kaltura-common/esm2015/lib/rxjs/operators/cancel-on-destroy.js
import { merge, Subject, Subscriber } from "rxjs";
const __ngOnDestroySource__ = Symbol("__ngOnDestroySource__");
const __ngOnDestroy__ = Symbol("__ngOnDestroy__");
export function cancelOnDestroy(instance, manualDestroy) {
    return (source) => source.lift(new CancelOnDestroyOperator(instance, manualDestroy));
}
class CancelOnDestroyOperator {
    constructor(instance, manualDestroy) {
        this.instance = instance;
        this.manualDestroy = manualDestroy;
        if (instance.ngOnDestroy) {
            if (!instance[__ngOnDestroySource__]) {
                instance[__ngOnDestroySource__] = new Subject();
                instance[__ngOnDestroy__] = instance.ngOnDestroy;
                instance.ngOnDestroy = function () {
                    this[__ngOnDestroy__].apply(this, arguments);
                    this[__ngOnDestroySource__].next();
                    this[__ngOnDestroySource__].complete();
                };
            }
        }
    }
    call(subscriber, source) {
        return source.subscribe(new CancelOnDestroySubscriber(subscriber, this.instance, this.manualDestroy));
    }
}
class CancelOnDestroySubscriber extends Subscriber {
    constructor(destination, _instance, manualDestroy) {
        super(destination);
        this._instance = _instance;
        this.manualDestroy = manualDestroy;
        const sources = manualDestroy
            ? merge(manualDestroy, _instance[__ngOnDestroySource__])
            : _instance[__ngOnDestroySource__].asObservable();
        this.add(sources.subscribe(() => {
            destination.unsubscribe();
        }));
    }
}
