// @kaltura-ng/kaltura-ui/esm5/lib/details-bar/detail-info.component.js
import { __decorate, __metadata } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
var DetailInfoComponent = /** @class */ (function () {
    function DetailInfoComponent() {
        this.toolTipAsHTML = true;
        this.separator = "|";
        this.maxItemWidth = 300;
        this.isLastItem = false;
        this.itemClick = new EventEmitter();
        this.isClickable = false;
    }
    DetailInfoComponent.prototype._setData = function (data) {
        this._data = data;
    };
    DetailInfoComponent.prototype.ngOnInit = function () {
        this.isClickable = this.itemClick.observers.length > 0;
    };
    DetailInfoComponent.prototype.onClick = function ($event) {
        if (this.isClickable) {
            this.itemClick.emit($event);
        }
    };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], DetailInfoComponent.prototype, "label", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], DetailInfoComponent.prototype, "value", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], DetailInfoComponent.prototype, "valueField", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], DetailInfoComponent.prototype, "link", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], DetailInfoComponent.prototype, "tooltip", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], DetailInfoComponent.prototype, "toolTipAsHTML", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], DetailInfoComponent.prototype, "iconStyle", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], DetailInfoComponent.prototype, "separator", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], DetailInfoComponent.prototype, "maxItemWidth", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], DetailInfoComponent.prototype, "isLastItem", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], DetailInfoComponent.prototype, "itemClick", void 0);
    DetailInfoComponent = __decorate([
        Component({
            selector: 'kDetailInfo',
            template: "<div class=\"kDetailInfo\" *ngIf=\"_data\">\n    <div class=\"kDetails\" [style.maxWidth]=\"maxItemWidth+'px'\">\n        <div *ngIf=\"iconStyle\" [kTooltip]=\"tooltip\" class=\"{{iconStyle}} kIcon\"></div>\n        <span class=\"kLabel\" *ngIf=\"label\">{{label}}</span>\n        <span *ngIf=\"!link && (value || valueField)\" [kTooltip]=\"tooltip\" [escape]=\"toolTipAsHTML\" [class.link]=\"isClickable\" (click)=\"onClick($event)\">{{value || _data[valueField]}}</span>\n        <a *ngIf=\"link\" [href]=\"link\" [target]=\"'_blank'\" [kTooltip]=\"tooltip\" class=\"link\">{{link}}</a>\n    </div>\n    <div *ngIf=!isLastItem class=\"kSeparator\">{{separator}}</div>\n</div> ",
            styles: [".kDetailInfo{display:flex;line-height:26px;height:26px}.kDetailInfo .kDetails{display:flex;overflow:hidden}.kDetailInfo .kDetails>a,.kDetailInfo .kDetails>div,.kDetailInfo .kDetails>span{flex:0 0 auto;line-height:26px}.kDetailInfo .kSeparator{padding-right:8px;padding-left:8px;flex:0 0 auto;color:#ccc}.kDetailInfo .kIcon{font-size:16px}.kDetailInfo .kLabel{margin-right:4px}.kDetailInfo .link{font-size:15px;color:#1b4a97;text-decoration:underline;cursor:pointer;text-overflow:ellipsis}.kDetailInfo .link:hover{color:#13356c}.kDetailInfo .link:active{color:#1b4a97}.kDetailInfo .link:disabled{cursor:none;color:#ccc}"]
        }),
        __metadata("design:paramtypes", [])
    ], DetailInfoComponent);
    return DetailInfoComponent;
}());
export { DetailInfoComponent };
