// @kaltura-ng/kaltura-ui/esm5/lib/tags/tag.component.js
import { __decorate, __metadata } from "tslib";
import { Component, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
var TagComponent = /** @class */ (function () {
    function TagComponent() {
        this.showRemove = true;
        this.onRemoved = new EventEmitter();
    }
    TagComponent.prototype.removeTag = function (tag) {
        this.onRemoved.emit(tag);
    };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], TagComponent.prototype, "label", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], TagComponent.prototype, "tooltip", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], TagComponent.prototype, "data", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], TagComponent.prototype, "disabled", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], TagComponent.prototype, "showRemove", void 0);
    __decorate([
        Input(),
        __metadata("design:type", TemplateRef)
    ], TagComponent.prototype, "template", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], TagComponent.prototype, "onRemoved", void 0);
    TagComponent = __decorate([
        Component({
            selector: 'kTag',
            template: "<div class=\"kTag\" [class.kTagDisabled]=\"disabled\" *ngIf=\"label\" [ngSwitch]=\"!!template\">\n\t<div *ngSwitchCase=\"true\">\n\t\t<!-- CUSTOM TEMPLATE -->\n\t\t<ng-container *ngTemplateOutlet=\"template;context:{item: { label: label, tooltip: tooltip, showRemove: showRemove, data: data }}\"></ng-container>\n\t</div>\n    <div class=\"kTagContent\" title=\"{{tooltip}}\" *ngSwitchCase=\"false\">\n        <span class=\"kTagLabel\">{{label}}</span>\n        <i *ngIf=\"showRemove\" class=\"kTagRemove kIconclose_small\" (click)=\"!disabled && removeTag(data)\"></i>\n    </div>\n</div>\n",
            styles: [":host{display:table-cell}.kTag{display:table-cell;height:24px;background-color:#00a784;border-radius:3px;color:#fff;font-size:14px;font-weight:700;padding-left:8px;padding-right:2px}.kTag.kTagDisabled,.kTag.kTagDisabled:hover{background-color:transparent;color:#ccc;border:1px solid #ccc;border-radius:3px}.kTagContent{display:flex;align-items:center}.kTagLabel{flex:0 0 auto;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:default}.kTagRemove{flex:0 0 auto;margin-top:2px;margin-left:8px;padding:5px;cursor:pointer;font-size:10px}.kTagRemove:hover{background-color:rgba(255,255,255,.3);border:0 solid rgba(255,255,255,.3);border-radius:3px}"]
        }),
        __metadata("design:paramtypes", [])
    ], TagComponent);
    return TagComponent;
}());
export { TagComponent };
