// @kaltura-ng/kaltura-common/esm2015/lib/operation-tag/operation-tag.module.js
var OperationTagModule_1;
import { __decorate, __metadata, __param } from "tslib";
import { Self, Optional, NgModule } from '@angular/core';
import { OperationTagStoreMediator } from "./operation-tag-store-mediator";
import { OperationTagManagerService } from "./operation-tag-manager.service";
let OperationTagModule = OperationTagModule_1 = class OperationTagModule {
    // Prevents angular from creating another instance for operationTagManagerService when defined as provider in more than one place
    // (Each component will get the same instance as in the module/component that called the operationTagManagerService.forRoot)
    constructor(operationTagManagerService) {
        if (operationTagManagerService) {
            OperationTagStoreMediator.register(operationTagManagerService);
        }
    }
    static forRoot() {
        return {
            ngModule: OperationTagModule_1,
            providers: [
                OperationTagManagerService
            ]
        };
    }
};
OperationTagModule.ctorParameters = () => [
    { type: OperationTagManagerService, decorators: [{ type: Self }, { type: Optional }] }
];
OperationTagModule = OperationTagModule_1 = __decorate([
    NgModule({
        imports: [],
        declarations: [],
        exports: [],
        providers: []
    }),
    __param(0, Self()), __param(0, Optional()),
    __metadata("design:paramtypes", [OperationTagManagerService])
], OperationTagModule);
export { OperationTagModule };
