// @kaltura-ng/kaltura-logger/bundles/kaltura-ng-kaltura-logger.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('jsnlog')) :
    typeof define === 'function' && define.amd ? define('@kaltura-ng/kaltura-logger', ['exports', '@angular/core', '@angular/common', 'jsnlog'], factory) :
    (global = global || self, factory((global['kaltura-ng'] = global['kaltura-ng'] || {}, global['kaltura-ng']['kaltura-logger'] = {}), global.ng.core, global.ng.common, global.jsnlog));
}(this, (function (exports, core, common, jsnlog) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    var KalturaLoggerRecordService = /** @class */ (function () {
        function KalturaLoggerRecordService() {
            this._recordLogs = false;
            this._logsBuffer = [];
        }
        Object.defineProperty(KalturaLoggerRecordService.prototype, "isRecordingLogs", {
            get: function () {
                return this._recordLogs;
            },
            enumerable: true,
            configurable: true
        });
        KalturaLoggerRecordService.prototype.startRecord = function () {
            this._recordLogs = true;
        };
        KalturaLoggerRecordService.prototype.getRecordedLogs = function () {
            var result = __spread(this._logsBuffer);
            this._recordLogs = false;
            this._logsBuffer = [];
            return result;
        };
        KalturaLoggerRecordService.prototype.addLogItemToBuffer = function (logItem) {
            if (this.isRecordingLogs) {
                this._logsBuffer.push(logItem);
            }
        };
        KalturaLoggerRecordService = __decorate([
            core.Injectable()
        ], KalturaLoggerRecordService);
        return KalturaLoggerRecordService;
    }());

    var KalturaLoggerName = new core.InjectionToken('kaltura-logger-name');
    var randomLoggerNameNumber = 1;
    var KalturaLogger = /** @class */ (function () {
        function KalturaLogger(name, parentLogger, _loggerRecordInterceptor) {
            this._loggerRecordInterceptor = _loggerRecordInterceptor;
            KalturaLogger_1.resetDefaultJSNLog();
            if (!name) {
                name = 'logger' + randomLoggerNameNumber;
                randomLoggerNameNumber++;
            }
            name = name.replace(/[.]/g, '_');
            this._name = parentLogger ? parentLogger.name + "." + name : name;
            this._logger = jsnlog.JL(this._name);
            this._logger.trace('logger created!');
        }
        KalturaLogger_1 = KalturaLogger;
        KalturaLogger.resetDefaultJSNLog = function () {
            if (!KalturaLogger_1.resetDefaultExecuted) {
                KalturaLogger_1.resetDefaultExecuted = true;
                var consoleAppender = jsnlog.JL.createConsoleAppender('consoleAppender');
                jsnlog.JL().setOptions({
                    appenders: [consoleAppender]
                });
            }
        };
        KalturaLogger.createLogger = function (loggerName) {
            return [
                KalturaLogger_1,
                {
                    provide: KalturaLoggerName, useValue: loggerName
                }
            ];
        };
        Object.defineProperty(KalturaLogger.prototype, "name", {
            get: function () {
                return this._name;
            },
            enumerable: true,
            configurable: true
        });
        KalturaLogger.prototype._addLogToBuffer = function (logItem) {
            if (this._loggerRecordInterceptor) {
                this._loggerRecordInterceptor.addLogItemToBuffer(logItem);
            }
        };
        KalturaLogger.prototype.startRecordingLogs = function () {
            if (this._loggerRecordInterceptor) {
                this._loggerRecordInterceptor.startRecord();
            }
        };
        KalturaLogger.prototype.getRecordedLogs = function () {
            if (this._loggerRecordInterceptor) {
                return this._loggerRecordInterceptor.getRecordedLogs();
            }
        };
        KalturaLogger.prototype.isValidLogLevel = function (level) {
            var validLogLevels = ['All', 'Trace', 'Debug', 'Info', 'Warn', 'Error', 'Fatal', 'Off'];
            return validLogLevels.indexOf(level) !== -1;
        };
        KalturaLogger.prototype.setOptions = function (options) {
            var level = undefined;
            if (this.isValidLogLevel(options.level) && jsnlog.JL) {
                var getLevelValue = jsnlog.JL["get" + options.level + "Level"];
                level = typeof getLevelValue === 'function' ? getLevelValue() : undefined;
            }
            jsnlog.JL().setOptions({ level: level });
        };
        KalturaLogger.prototype.subLogger = function (name) {
            return new KalturaLogger_1(name, this, this._loggerRecordInterceptor);
        };
        KalturaLogger.prototype.ngOnDestroy = function () {
            this._logger.debug('logger destroyed');
            delete this._logger;
        };
        KalturaLogger.prototype._createLogObject = function (level, message, context) {
            this._addLogToBuffer({ level: level, message: message, context: context });
            return context ? Object.assign({ message: message, level: level }, context) : message;
        };
        KalturaLogger.prototype.trace = function (message, context) {
            var _this = this;
            if (context && typeof context === 'function') {
                this._logger.trace(function () { return _this._createLogObject('trace', message, context()); });
            }
            else {
                this._logger.trace(this._createLogObject('trace', message, context));
            }
        };
        KalturaLogger.prototype.debug = function (message, context) {
            var _this = this;
            if (context && typeof context === 'function') {
                this._logger.debug(function () { return _this._createLogObject('debug', message, context()); });
            }
            else {
                this._logger.debug(this._createLogObject('debug', message, context));
            }
        };
        KalturaLogger.prototype.info = function (message, context) {
            var _this = this;
            if (context && typeof context === 'function') {
                this._logger.info(function () { return _this._createLogObject('info', message, context()); });
            }
            else {
                this._logger.info(this._createLogObject('info', message, context));
            }
        };
        KalturaLogger.prototype.warn = function (message, context) {
            var _this = this;
            if (context && typeof context === 'function') {
                this._logger.warn(function () { return _this._createLogObject('warn', message, context()); });
            }
            else {
                this._logger.warn(this._createLogObject('warn', message, context));
            }
        };
        KalturaLogger.prototype.error = function (message, context) {
            this._logger.error(this._createLogObject('error', message, context));
        };
        KalturaLogger.prototype.fatal = function (message, context) {
            if (context && context instanceof Error) {
                this._logger.fatalException(message, context);
            }
            else {
                this._logger.fatal(this._createLogObject('fatal', message, context));
            }
        };
        var KalturaLogger_1;
        KalturaLogger.resetDefaultExecuted = false;
        KalturaLogger.ctorParameters = function () { return [
            { type: String, decorators: [{ type: core.Inject, args: [KalturaLoggerName,] }, { type: core.Optional }, { type: core.Self }] },
            { type: KalturaLogger, decorators: [{ type: core.SkipSelf }, { type: core.Optional }] },
            { type: KalturaLoggerRecordService, decorators: [{ type: core.Optional }] }
        ]; };
        KalturaLogger = KalturaLogger_1 = __decorate([
            core.Injectable(),
            __param(0, core.Inject(KalturaLoggerName)), __param(0, core.Optional()), __param(0, core.Self()),
            __param(1, core.SkipSelf()), __param(1, core.Optional()),
            __param(2, core.Optional()),
            __metadata("design:paramtypes", [String, KalturaLogger,
                KalturaLoggerRecordService])
        ], KalturaLogger);
        return KalturaLogger;
    }());

    if (window && window.onerror) {
        window.onerror = null;
    }
    var KalturaLoggerModule = /** @class */ (function () {
        function KalturaLoggerModule() {
        }
        KalturaLoggerModule_1 = KalturaLoggerModule;
        KalturaLoggerModule.forRoot = function (name) {
            return {
                ngModule: KalturaLoggerModule_1,
                providers: [
                    KalturaLogger,
                    {
                        provide: KalturaLoggerName, useValue: name
                    },
                    KalturaLoggerRecordService
                ]
            };
        };
        var KalturaLoggerModule_1;
        KalturaLoggerModule = KalturaLoggerModule_1 = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule
                ],
                declarations: [],
                exports: [],
                providers: []
            })
        ], KalturaLoggerModule);
        return KalturaLoggerModule;
    }());

    exports.KalturaLogger = KalturaLogger;
    exports.KalturaLoggerModule = KalturaLoggerModule;
    exports.KalturaLoggerName = KalturaLoggerName;
    exports.ɵa = KalturaLoggerRecordService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
